### ISSUE TITLE
Sender may busy loop when connection is in CONNECTING state

### ISSUE THREAD

**Issue by rikonen** [dpkp/kafka-python#1907]
The following may happen:

1. kafka-python successfully fetches metadata for cluster indicating that leader for topic partition `X` is node `Y`
2. Client application starts producing data, including some that goes to topic partition `X`
3. kafka-python successfully connects to `Y` and starts sending batches to node `Y` as soon as they're completed
4. kafka-python loses connection to `Y`. Client still continues producing data
5. kafka-python tries establishing connection to `Y` but the connection doesn't proceed beyond `CONNECTING` state in a while
6. New batch going to node `Y` completes.
7. `Sender.run_once()` gets node Y from `RecordAccumulator.ready()` but `self._client.is_ready()` returns `False` because there's no working connection. `run_once` discards node `Y` from the list of nodes to process and calls `self._client.connection_delay()` to determine how long to block in `self._client.poll()`. Because the connection to Y is in state `CONNECTING`, `self._client.connection_delay()` returns 0.
8. `run_once()` calls `self._client.poll()` with timeout zero, causing `poll()` to exit practically immediately.
9. `run()` calls `run_once()` again with no delay and sequence continues from step 7: kafka-python keeps on busy looping until connection to Y is successfully established, which may take very long time if the broker has some problem that prevents new connections from being established.

Log extract showing this behavior below:

```
2019-09-23T15:54:50.542310  hostname  kafka.conn            kafka-python-producer-2-network-thread INFO      <BrokerConnection node_id=71 host=10.0.1.100:13737 <connecting> [IPv4 ('10.0.1.100', 13737)]>: connecting to 10.0.1.100:13737 [('10.0.1.100', 13737) IPv4]
2019-09-23T15:54:59.819240  hostname  kafka.producer.sender  kafka-python-producer-2-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
2019-09-23T15:54:59.849336  hostname  kafka.producer.sender  kafka-python-producer-1-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
2019-09-23T15:54:59.865445  hostname  kafka.producer.sender  kafka-python-producer-1-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
2019-09-23T15:54:59.868538  hostname  kafka.producer.sender  kafka-python-producer-2-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
2019-09-23T15:54:59.869473  hostname  kafka.producer.sender  kafka-python-producer-2-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
2019-09-23T15:54:59.869154  hostname  kafka.producer.sender  kafka-python-producer-1-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
... same repeats a lot ...
2019-09-23T15:55:00.964323  hostname  kafka.producer.sender  kafka-python-producer-2-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
2019-09-23T15:55:03.119806  hostname  kafka.client          kafka-python-producer-1-network-thread WARNING   Node 71 connection failed -- refreshing metadata
2019-09-23T15:55:03.119478  hostname  kafka.conn            kafka-python-producer-1-network-thread ERROR     Connection attempt to <BrokerConnection node_id=71 host=10.0.1.100:13737 <handshake> [IPv4 ('10.0.1.100', 13737)]> timed out
...
2019-09-23T15:55:03.136644  hostname  kafka.conn            kafka-python-producer-1-network-thread INFO      <BrokerConnection node_id=71 host=10.0.1.100:13737 <connecting> [IPv4 ('10.0.1.100', 13737)]>: connecting to 10.0.1.100:13737 [('10.0.1.100', 13737) IPv4]
2019-09-23T15:55:03.137528  hostname  kafka.producer.record_accumulator  kafka-python-producer-1-network-thread WARNING   Expired 1 batches in accumulator
2019-09-23T15:55:03.137289  hostname  kafka.producer.record_accumulator  kafka-python-producer-1-network-thread WARNING   Produced messages to topic-partition TopicPartition(topic='topic-name', partition=1) with base offset -1 and error KafkaTimeoutError: Batch for TopicPartition(topic='topic-name', partition=1) containing 40 record(s) expired: 30 seconds have passed since last append.
2019-09-23T15:55:03.174292  hostname  kafka.producer.record_accumulator  kafka-python-producer-1-network-thread WARNING   Expired 1 batches in accumulator
2019-09-23T15:55:03.174028  hostname  kafka.producer.record_accumulator  kafka-python-producer-1-network-thread WARNING   Produced messages to topic-partition TopicPartition(topic='topic-name', partition=1) with base offset -1 and error KafkaTimeoutError: Batch for TopicPartition(topic='topic-name', partition=1) containing 49 record(s) expired: 30 seconds have passed since last append.
2019-09-23T15:55:03.192773  hostname  kafka.producer.record_accumulator  kafka-python-producer-1-network-thread WARNING   Produced messages to topic-partition TopicPartition(topic='topic-name', partition=1) with base offset -1 and error KafkaTimeoutError: Batch for TopicPartition(topic='topic-name', partition=1) containing 42 record(s) expired: 30 seconds have passed since last append.
2019-09-23T15:55:03.193092  hostname  kafka.producer.record_accumulator  kafka-python-producer-1-network-thread WARNING   Expired 1 batches in accumulator
... same repeats but not that much ...
2019-09-23T15:55:20.548887  hostname  kafka.conn            kafka-python-producer-2-network-thread INFO      <BrokerConnection node_id=71 host=10.0.1.100:13737 <handshake> [IPv4 ('10.0.1.100', 13737)]>: Closing connection. KafkaConnectionError: timeout
2019-09-23T15:55:20.548724  hostname  kafka.conn            kafka-python-producer-2-network-thread ERROR     Connection attempt to <BrokerConnection node_id=71 host=10.0.1.100:13737 <handshake> [IPv4 ('10.0.1.100', 13737)]> timed out
2019-09-23T15:55:20.549080  hostname  kafka.client          kafka-python-producer-2-network-thread WARNING   Node 71 connection failed -- refreshing metadata
...
2019-09-23T15:55:20.615466  hostname  kafka.conn            kafka-python-producer-2-network-thread INFO      <BrokerConnection node_id=71 host=10.0.1.100:13737 <connecting> [IPv4 ('10.0.1.100', 13737)]>: connecting to 10.0.1.100:13737 [('10.0.1.100', 13737) IPv4]
...
2019-09-23T15:55:29.819565  hostname  kafka.producer.sender  kafka-python-producer-1-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
2019-09-23T15:55:29.819876  hostname  kafka.producer.sender  kafka-python-producer-1-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
2019-09-23T15:55:29.820507  hostname  kafka.producer.sender  kafka-python-producer-1-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
2019-09-23T15:55:29.820832  hostname  kafka.producer.sender  kafka-python-producer-2-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
... same repeats a lot ...
2019-09-23T15:55:30.405083  hostname  kafka.producer.sender  kafka-python-producer-2-network-thread DEBUG     Node 71 not ready; delaying produce of accumulated batch
2019-09-23T15:55:33.136181  hostname  kafka.conn            kafka-python-producer-1-network-thread ERROR     Connection attempt to <BrokerConnection node_id=71 host=10.0.1.100:13737 <handshake> [IPv4 ('10.0.1.100', 13737)]> timed out
```

**Comment by dpkp** [comment]
Thanks for the issue report. I think we can actually just return float('inf') and rely on the network selector to notify when connections are complete via an EVENT_WRITE. I'll put up a PR to handle this.

**Comment by dpkp** [comment]
Fixed with #1909
